executor/cte.go

```diff
@@ -0,0 +1,432 @@
+
+	seedExec      Executor
+	recursiveExec Executor
+
+	// `resTbl` and `iterInTbl` are shared by all CTEExec which reference to same the CTE.
+	// `iterInTbl` is also shared by CTETableReaderExec.
+	resTbl     cteutil.Storage
+	iterInTbl  cteutil.Storage
+	iterOutTbl cteutil.Storage
+
+	resHashTbl    baseHashTable
+	iterInHashTbl baseHashTable
+
+	// Index of chunk to read from `resTbl`.
+	chkIdx int
+
+	// UNION ALL or UNION DISTINCT.
+	isDistinct bool
+	curIter    int
+}
+
+// Open implements the Executor interface.
+func (e *CTEExec) Open(ctx context.Context) (err error) {
+	e.reset()
+	if err := e.baseExecutor.Open(ctx); err != nil {
+		return err
+	}
+
+	if e.seedExec == nil {
+		return errors.New("seedExec for CTEExec is nil")
+	}
+	if err = e.seedExec.Open(ctx); err != nil {
+		return err
+	}
+
+	if e.recursiveExec != nil {
+		if err = e.recursiveExec.Open(ctx); err != nil {
+			return err
+		}
+		recursiveTypes := e.recursiveExec.base().retFieldTypes
+		e.iterOutTbl = cteutil.NewStorageRC(recursiveTypes, e.maxChunkSize)
+		if err = e.iterOutTbl.OpenAndRef(); err != nil {
+			return err
+		}
+
+		setupCTEStorageTracker(e.iterOutTbl, e.ctx)
+	}
+
+	if e.isDistinct {
+		e.resHashTbl = newConcurrentMapHashTable()
+		e.iterInHashTbl = newConcurrentMapHashTable()
+	}
+	return nil
+}
+
+// Next implements the Executor interface.
+func (e *CTEExec) Next(ctx context.Context, req *chunk.Chunk) (err error) {
+	req.Reset()
+	e.resTbl.Lock()
+	defer func() {
+		e.resTbl.Unlock()
+	}()
+	if !e.resTbl.Done() {
+		setupCTEStorageTracker(e.resTbl, e.ctx)
+		setupCTEStorageTracker(e.iterInTbl, e.ctx)
+
+		if err = e.computeSeedPart(ctx); err != nil {
+			// Don't put it in defer.
+			// Because it should be called only when the filling process is not completed.
+			if err1 := e.reopenTbls(); err1 != nil {
+				return err1
+			}
+			return err
+		}
+		if err = e.computeRecursivePart(ctx); err != nil {
+			if err1 := e.reopenTbls(); err1 != nil {
+				return err1
+			}
+			return err
+		}
+	}
+	e.resTbl.SetDone()
+
+	if e.chkIdx < e.resTbl.NumChunks() {
+		res, err := e.resTbl.GetChunk(e.chkIdx)
+		if err != nil {
+			return err
+		}
+		// Need to copy chunk to make sure upper operator will not change chunk in resTbl.
+		req.SwapColumns(res.CopyConstruct())
+		e.chkIdx++
+	}
+	return nil
+}
+
+// Close implements the Executor interface.
+func (e *CTEExec) Close() (err error) {
+	e.reset()
+	if err = e.seedExec.Close(); err != nil {
+		return err
+	}
+	if e.recursiveExec != nil {
+		if err = e.recursiveExec.Close(); err != nil {
+			return err
+		}
+	}
+
+	// `iterInTbl` and `resTbl` are shared by multiple operators,
+	// so will be closed when the SQL finishes.
+	if err = e.iterOutTbl.DerefAndClose(); err != nil {
+		return err
+	}
+	return e.baseExecutor.Close()
+}
+
+func (e *CTEExec) computeSeedPart(ctx context.Context) (err error) {
+	e.curIter = 0
+	e.iterInTbl.SetIter(e.curIter)
+	if e.curIter >= e.ctx.GetSessionVars().CTEMaxRecursionDepth {
```

Comment:
Why do we need to check this for the seed part?

Reply:
cte_max_recursion_depth == 0

Reply:
Do we have a test for this case?

Reply:
Added

Reply:
Note that here is the seed part. So you need some cases for nonrecursive CTE.

Reply:
done

Reply:
Cases added in TestCTEMaxRecursionDepth()